lib/facter/settings.rb

```diff
@@ -51,3 +51,16 @@
     File.expand_path(File.join(Puppet.settings['confdir'],'../../mcollective/etc'))
   end
 end
+
+Facter.add('puppet_agent_environment') do
+  setcode do
+    # There has to be a more graceful way to accomplish this, but I cannot seem to find a way
+    # via the existing ruby library in puppet... I think face action :config is worth looking at
+    # but have not been able to make it work, so this will have to do.
+    #
+    # This fact gets the puppet.conf config environment setting independent of the enc override or 
+    # the cli --environment override. The goal is to specify the agent environment to put into the 
+    # puppet.conf file by the msi installer without accidentally overriding it.
+    Facter::Core::Execution.exec('puppet config print --section agent environment');
```

Comment:
I think the following should work (from https://github.com/puppetlabs/puppet/blob/master/lib/puppet/face/config.rb#L57-L84)
```
values_from_the_selected_section = Puppet.settings.values(nil, :agent)

loader_settings = {
  :environmentpath => values_from_the_selected_section.interpolate(:environmentpath),
  :basemodulepath => values_from_the_selected_section.interpolate(:basemodulepath),
}

Puppet.override(Puppet.base_context(loader_settings),
    "New environment loaders generated from the requested section.") do
  # And now we can lookup values that include those from environments configured from
  # the requested section
  values = Puppet.settings.values(Puppet[:environment].to_sym, :agent)
  value = values.interpolate(:environment)
end
```